src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -34,41 +36,170 @@ public static class TagHelperDescriptorFactory
+            new[] { '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'' });
+
+        private static IEnumerable<TargetElementAttribute> GetValidTargetElementAttributes(
+            TypeInfo typeInfo,
+            ParserErrorSink errorSink)
+        {
+            var targetElementAttributes = typeInfo.GetCustomAttributes<TargetElementAttribute>(inherit: false);
+
+            return targetElementAttributes.Where(attribute => ValidTargetElementAttributeNames(attribute, errorSink));
+        }
+
+        private static IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
+            TypeInfo typeInfo,
+            string assemblyName,
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
+            IEnumerable<TargetElementAttribute> targetElementAttributes)
+        {
+            var typeName = typeInfo.FullName;
+
+            // If there isn't an attribute specifying the tag name derive it from the name
+            if (!targetElementAttributes.Any())
+            {
+                var name = typeInfo.Name;
+
+                if (name.EndsWith(TagHelperNameEnding, StringComparison.OrdinalIgnoreCase))
+                {
+                    name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
+                }
+
+                return BuildTagHelperDescriptors(
+                    tagNames: new[] { ToHtmlCase(name) },
+                    typeName: typeName,
+                    assemblyName: assemblyName,
+                    attributeDescriptors: attributeDescriptors,
+                    requiredAttributes: Enumerable.Empty<string>());
+            }
+
+            return targetElementAttributes.SelectMany(
+                attribute => BuildTagHelperDescriptors(typeName, assemblyName, attributeDescriptors, attribute));
+        }
+
+        private static IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
+            string typeName,
+            string assemblyName,
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
+            TargetElementAttribute targetElementAttribute)
+        {
+            var tagNames = GetCommaSeparatedValues(targetElementAttribute.Tag);
+            var requiredAttributes = GetCommaSeparatedValues(targetElementAttribute.Attributes);
+
+            return BuildTagHelperDescriptors(
+                tagNames,
+                typeName,
+                assemblyName,
+                attributeDescriptors,
+                requiredAttributes);
+        }
 
+        private static IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
+            IEnumerable<string> tagNames,
+            string typeName,
+            string assemblyName,
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
+            IEnumerable<string> requiredAttributes)
+        {
             return tagNames.Select(tagName =>
                 new TagHelperDescriptor(
```

Comment:
I like having the method because it's 1 location to provide `prefix: string.Empty`. Going to transition it to only take a single tag name however and modify name to be singular.